### PR TITLE
tutorial: Add draft version for Mistral Finetuning tutorial

### PR DESCRIPTION
**Pull request by tomaarsen**
Hello!

# Description

This is a (draft) PR for the Mistral finetuning tutorial to produce a chat-style LLM using a subset of the OIG dataset. The code is complete, but it still has to be explained and cleaned up a bit.

**Type of change**

- [x] Documentation update

**How Has This Been Tested**

- [ ] `sphinx-autobuild` (read [Developer Documentation](https://docs.argilla.io/en/latest/community/developer_docs.html#building-the-documentation) for more details)

**Checklist**

- [ ] I added relevant documentation
- [ ] I followed the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See https://keepachangelog.com/)

---

- Tom Aarsen

### CODE REVIEW

File: docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb [hunk 1]
```diff
@@ -0,0 +1,602 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Ⓜ️ Finetuning LLMs: Supervised Finetuning on Mistral 7B\n",
```
Comment (davidberenstein1957) [inline docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb]: maybe emphasize it is chat based in the title too?

File: docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb [hunk 2]
```diff
@@ -0,0 +1,602 @@
...
+    "\n",
+    "We recommend to set up the questions based on your goals and values for the annotation process. When preparing the training data, all of the annotations will be available, allowing you to carefully curate your training data. For example, based on your annotations you will be able to disregard samples that are (frequently) marked as harmful or low quality, or you can use the annotator-provided suggestions to responses or prompts instead of the existing ones.\n",
+    "Furthermore, in some situations you may only have prompts to begin with, and you can use Argilla to request your annotators to provide responses. You can then use these in your training data.\n",
+    "\n",
+    "### For this tutorial\n",
+    "\n",
+    "For the purposes of this tutorial, we will use a \"prompt\" field with all chat history and the current prompt, as well as a \"response\" field with the response. Additionally, we introduce a \"background\". For the questions we will use one simple `LabelQuestion` that asks the annotator whether the response is `\"Good\"` or `\"Bad\"`."
```
Comment (davidberenstein1957) [inline docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb]: background? maybe chat_history is more explicit?

File: docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb [hunk 3]
```diff
@@ -0,0 +1,602 @@
...
+    "```\n",
+    "Let's break this down:\n",
```
Comment (davidberenstein1957) [inline docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb]: maybe mention that this differs per model?

File: docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb [hunk 4]
```diff
@@ -0,0 +1,602 @@
...
+   "source": [
+    "![training-llm-mistral-sft](../../_static/tutorials/training-llm-mistral-sft/training-llm-mistral-sft.png)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## (Optional) Annotate instruction samples\n",
+    "If you are using your own proprietary data, data from an unreliable source, or require on your annotators to provide responses, then you must perform data annotation. However, for the purposes of this tutorial, we will assume that all data is high quality and skip this step."
```
Comment (davidberenstein1957) [inline docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb]: maybe add a reference to the docs/video from Natalia here?

File: docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb [hunk 5]
```diff
@@ -0,0 +1,602 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Set up the ArgillaTrainer for Supervised Finetuning\n",
+    "Next, we have to set up the [ArgillaTrainer](https://docs.argilla.io/en/latest/practical_guides/fine_tune.html#supervised-finetuning). First off, we will load the Mistral 7B `model` and `tokenizer`.\n"
```
Comment (davidberenstein1957) [inline docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb]: maybe add some context about the configs used when loadng the model? Why do you change the pad token etc

File: docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb [hunk 6]
```diff
@@ -0,0 +1,602 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "\n",
+    "\n",
+    "from tokenizers.processors import TemplateProcessing\n",
```
Comment (davidberenstein1957) [inline docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb]: some context can be added here too

File: docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb [hunk 7]
```diff
@@ -0,0 +1,602 @@
...
+    "        super().__init__()\n",
+    "        self.stopping_input_ids = stopping_input_ids\n",
+    "\n",
+    "    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> bool:\n",
+    "        for stopping_input_ids in self.stopping_input_ids:\n",
+    "            length = len(stopping_input_ids)\n",
+    "            if input_ids[0, -length:].tolist() == stopping_input_ids:\n",
+    "                return True\n",
+    "\n",
+    "stopping_criteria = StoppingCriteriaList([StoppingCriteriaTokens(tokenizer_without_eos([\"<bot>:\", \"<human>:\"]).input_ids)])\n",
```
Comment (davidberenstein1957) [inline docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb]: I would highlight the stopping criteria too

File: docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb (hunk 1 above)
Comment (tomaarsen) [reply docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb]: Great idea, handled in 3d221adbae61bcf23ca8aee740a31f678a6df230

File: docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb (hunk 2 above)
Comment (tomaarsen) [reply docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb]: This is indeed a bit vague. The "chat history" is all inside of the prompt, the "background" is just some additional bit of text that *some* samples of the OIG dataset have before the whole `<human>:` `<bot>:` steps. The loaded data is a good example, actually.
image


I elaborate a bit better here: f4b91ad2817d9a77673cac7de680020e5f47ec73

File: docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb (hunk 3 above)
Comment (tomaarsen) [reply docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb]: It doesn't, this is just specific to the data, not the model.

And for the model we make our own template, which you can always do when finetuning a base model.

File: docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb (hunk 4 above)
Comment (tomaarsen) [reply docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb]: Which reference do you think is best?
https://docs.argilla.io/en/latest/practical_guides/annotate_dataset.html perhaps?

File: docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb (hunk 5 above)
Comment (tomaarsen) [reply docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb]: Covered in 06e5319750a3f4f5947bc6ed6675a9c9d5437afd

File: docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb (hunk 6 above)
Comment (tomaarsen) [reply docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb]: This has been removed, `TemplateProcessing` was only necessary before I changed up the tokenization approach.

File: docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb (hunk 7 above)
Comment (tomaarsen) [reply docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb]: This has also been removed - `StoppingCriteriaList` was too inconsistent of an approach to stop with, the current way with EOS is much more consistent. This was just leftover from an earlier version

**Comment by tomaarsen** [comment]
TODO: Add links to this tutorial from elsewhere in the docs, toctrees etc.

File: docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb [hunk 8]
```diff
@@ -0,0 +1,946 @@
...
+    "\n",
+    "We recommend to set up the questions based on your goals and values for the annotation process. When preparing the training data, all of the annotations will be available, allowing you to carefully curate your training data. For example, based on your annotations you will be able to disregard samples that are (frequently) marked as harmful or low quality, or you can use the annotator-provided suggestions to responses or prompts instead of the existing ones.\n",
+    "Furthermore, in some situations you may only have prompts to begin with, and you can use Argilla to request your annotators to provide responses. You can then use these in your training data.\n",
+    "\n",
+    "### For this tutorial\n",
+    "\n",
+    "For the purposes of this tutorial, we will use a \"prompt\" field with all chat history and the current prompt, as well as a \"response\" field with the response. Additionally, we introduce a \"background\" text field, which is sometimes used to provide additional background information prior to the user prompt. For the questions we will use one simple `LabelQuestion` that asks the annotator whether the response is `\"Good\"` or `\"Bad\"`."
```
Comment (davidberenstein1957) [inline docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb]: maybe we can provide an example for background because it is an uncommon thing? or at least to me, and the explanation is not too clear about a concise example yet.

Comment (davidberenstein1957) [reply docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb]: maybe we can mention that an example of background is shown underneath

File: docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb (hunk 3 above)
Comment (davidberenstein1957) [reply docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb]: Yes, I think it is a two-fold thing which also includes configuring model token format later on in the process. I think it might be nice to briefly mention the complexities/background of this, similar to the reason why they added the tokenizer implementation to transformers right?

File: docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb (hunk 4 above)
Comment (davidberenstein1957) [reply docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb]: Yes that would be good but maybe https://docs.argilla.io/en/latest/conceptual_guides/llm/llm.html is better from a conceptual point of view?

File: docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb [hunk 9]
```diff
@@ -0,0 +1,946 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Set up the ArgillaTrainer for Supervised Finetuning\n",
+    "### Model & Tokenizer\n",
+    "Next, we have to set up the [ArgillaTrainer](https://docs.argilla.io/en/latest/practical_guides/fine_tune.html#supervised-finetuning). First off, we will load the Mistral 7B `model` and `tokenizer`. We'll load the model using `float16` to improve the memory usage and efficiency, and `device_map=\"auto\"` automatically picks the best device to load the model on. For example, this will prioritize your GPU before your CPU.\n",
+    "\n",
+    "Furthermore, setting the `pad_token_id` to `eos_token_id` is required for open-end generation. If you don't define `pad_token_id`, it is often set to `eos_token_id` already, but you will be given warnings that you should do it yourself."
```
Comment (davidberenstein1957) [inline docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb]: this is part of the context I referenced above w.r.t. markers/tokens.

File: docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb [hunk 10]
```diff
@@ -0,0 +1,946 @@
...
+   "source": [
+    "This seems to match our desired format exactly!"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Data Collation\n",
+    "Next, we want to verify that this text is tokenized and collated correctly. In particular the data collation can be quite important, as some data collators will replace all special tokens (e.g. padding, BOS and EOS tokens) with a label of -100, which is the label that is ignored during training. If this is the case, then the model will not learn to produce the EOS token. Consequently, the model will always respond with endless tokens - far from ideal.\n",
```
Comment (davidberenstein1957) [inline docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb]: This is indeed part of the context I mentioned above but maybe a general intro would be nice.

**Comment by tomaarsen** [comment]
@davidberenstein1957 Could you verify if it's correctly placed in the docs? It's extremely confusing to figure out what I need to edit. I used [this](https://github.com/argilla-io/argilla/pull/3592/files#diff-5b12fd5218b41418f3093fb1ba841630be798547ac47b52017cee46c9158a196) as a guide, but it doesn't make any sense, do we even still show libraries, steps, tasks, techniques anymore?

File: docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb (hunk 8 above)
Comment (tomaarsen) [reply docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb]: Resolved in 699beef4e941635d0402e1cb838709992f2b2f04

**Comment by davidberenstein1957** [comment]
@tomaarsen that is correct and it looks like it has been added well :) For context, I try to emphasize the background a bit better because I want to have clear end-2-end example blog-like tutorials for each basic task we support with the FeedbackDataset.

File: docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb (hunk 4 above)
Comment (tomaarsen) [reply docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb]: Sounds good, solved in c53ac20136bfdf6c5ecda13067b4dbab25e4c970

File: docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb (hunk 10 above)
Comment (tomaarsen) [reply docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb]: Handled in 2838cccc43d0e713b041c8c876df8211cc1f5999, I rephrased this slightly to make the intro a bit more smooth

**Comment by davidberenstein1957** [comment]
@tomaarsen bye bye ❤️ -> ❤️‍🩹
